File: ovstorage-core/examples/python/03_load_a_plugin.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
"""Load the core plugin explicitly and put its Router over file."""

from __future__ import annotations

import argparse
import asyncio
import tempfile
from pathlib import Path

import ovstorage
from ovstorage.file import FileBackend
from ovstorage.router import Router

from _common import file_connection_request, plugin_registry


async def run(plugin_dir: str | None) -> None:
    with tempfile.TemporaryDirectory(prefix="ovstorage-03-") as directory:
        root = Path(directory)

        # Unlike the built-in file backend, Router comes from the core plugin.
        # An explicit registry loads that shared library and makes its Layer
        # factories available while this Stack is built.
        registry = plugin_registry(plugin_dir, "core")

        # Requests enter `routes`, whose ordered children are the candidate
        # backends. The router compares each child's declared address roots and
        # sends this example's file:// URL to `files`.
        #
        #     routes (root/router)
        #       └── files (backend + connection)
        storage = await (
            ovstorage.Stack(root="routes")
            .with_registry(registry)
            .router(Router("routes", ["files"]))
            .backend(FileBackend("files"))
            .connection("files", file_connection_request(root, "local files"))
            .build()
        )

        # Callers use the same storage API regardless of how many routing or
        # wrapper Layers sit between the Stack root and the selected backend.
        address = (root / "routed.txt").as_uri()
        await storage.write(address, b"the core plugin routed this request\n")
        data, _info = await storage.read_bytes(address)
        print(data.decode().rstrip())


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--plugin-dir")
    args = parser.parse_args()
    asyncio.run(run(args.plugin_dir))


if __name__ == "__main__":
    main()

File: ovstorage-c-source/examples/tutorial/03_load_a_plugin.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

#include "common.hpp"

#include <cstddef>
#include <span>
#include <string_view>

int main(int argc, char** argv)
{
    if (argc != 2) {
        std::cerr << "usage: " << argv[0] << " PLUGIN_DIR\n";
        return 2;
    }
    tutorial::Context context;
    // Loading registers the core plugin's Layer factories and keeps the shared
    // library alive for every Layer handle built from it.
    if (!context.load(argv[1], "core")) return 1;

    tutorial::TempDirectory work("ovstorage-cpp-03");
    if (!work) return 1;
    const auto& directory = work.path();
    ovstorage::Stack stack;
    if (!tutorial::add_file(stack, context, "files", directory)) return 1;
    // Layer names connect the graph. Requests enter `routes`; Router compares
    // the address against its children's declared roots and selects `files`.
    //
    //     routes (root/router from the core plugin)
    //       `-- files (backend + connection)
    if (!tutorial::ok(
            "Stack::add_layer(router)",
            stack.add_layer(context.registry, "routes", "router"))) return 1;
    if (!tutorial::ok(
            "Stack::set_children",
            stack.set_children("routes", {"files"}))) return 1;
    if (!tutorial::ok("Stack::set_root", stack.set_root("routes"))) return 1;

    // build() validates all names and edges before it instantiates the graph.
    auto built = ovstorage::sync_wait(stack.build());
    if (!tutorial::ok("Stack::build", built)) return 1;
    auto storage = std::move(built).value();
    const std::string address =
        tutorial::file_root(directory) + "routed.txt";
    constexpr std::string_view message = "routed through the core plugin\n";
    // The caller still uses the ordinary storage API; routing is internal to
    // the configured Stack rather than selected at each call site.
    auto written = ovstorage::sync_wait(storage.write(
        address,
        std::as_bytes(std::span(message.data(), message.size()))));
    if (!tutorial::ok("write through routes", written)) return 1;
    auto read = ovstorage::sync_wait(storage.read_bytes(address));
    if (!tutorial::ok("read through routes", read)) return 1;
    if (read.value().first.string() != message) {
        std::cerr << "routed read returned the wrong bytes\n";
        return 1;
    }
    std::cout << read.value().first.string();
    return 0;
}
